### ISSUE TITLE
macOs and deprecation of OpenGL and OpenCL.

### ISSUE THREAD

**Issue by KelSolaar** [vispy/vispy#1484]
I suppose most of you have read about it: https://developer.apple.com/macos/whats-new/#deprecationofopenglandopencl

I wanted to know if there were discussions about the consequences this would have for Vispy on macOs.

Cheers,

Thomas

**Comment by sumanthratna** [comment]
@djhoese was correct about Metal; if you navigate to [this link](https://developer.apple.com/library/archive/documentation/GraphicsImaging/Conceptual/OpenGL-MacProgGuide/opengl_intro/opengl_intro.html) you'll see a banner that says Metal is now recommended. 

As for the urgency of this issue, Open˝GL has now become a problem for me (macOS Big Sur). See https://github.com/napari/napari/issues/1393 for diagnostic info. 

Hopefully this isn't beyond the scope of vispy but it'd be great if vispy could fall back to Metal on macOS systems when necessary.

**Comment by jni** [comment]
@sumanthratna I think (a) it's going to be a while before VisPy builds on top of Metal, though we would of course appreciate help there, but (b) that appears to still be a beta/configuration issue rather than the imminent destruction of OpenGL on Mac, based on [this tweet](https://twitter.com/colincornaby/status/1275153748348682240).

**Comment by djhoese** [comment]
> it'd be great if vispy could fall back to Metal on macOS systems when necessary.

Yes this would be great :wink: but isn't something simple like "import this backend instead of this one". Running using Metal (as far as I know) would be a complete rewrite of vispy. Although vispy separates the drawing instructions from the actual drawing (at least usually), these instructions are all based on OpenGL concepts. I'm not sure these concepts and workflows can be translated easily to Metal. It will also be difficult to do this work if there isn't a well maintained python-metal wrapper.

Another more likely option is Vulkan instead of Metal. This is something @almarklein has brought up in the past too (a vulkan-based vispy-like library).

**Comment by Russell-Jones-OxPhys** [comment]
Zink and MoltenVK are both open source. I don't know how useable/deployable they are. My impression is that MoltenVK is pretty complete and Zink is experimental and not getting much attention.
https://www.collabora.com/news-and-blog/blog/2018/10/31/introducing-zink-opengl-implementation-vulkan/ (OpenGL on Vulkan)
https://moltengl.com/moltenvk/ (Vulkan for macOS)

**Comment by sumanthratna** [comment]
MoltenVK looks promising—as for what @djhoese mentioned, [https://github.com/DiligentGraphics/DiligentEngine](DiligentGraphics/DiligentEngine) might help with not messing up the codebase (although I do understand if this is something vispy will never implement). 

Hopefully, it doesn't come down to this and we can stick to OpenGL for now. I'll try to debug a little more and see if we can solve this without switching backends.

**Comment by almarklein** [comment]
> Another more likely option is Vulkan instead of Metal. This is something @almarklein has brought up in the past too (a vulkan-based vispy-like library).

I did, but things have evolved ... in a good way. We're building on top of [wgpu](https://github.com/gfx-rs/wgpu), a higher level GPU API than Vulkan and Metal (but based on them, under the hood). It's build by Mozilla and is basically the Rust implementation of the upcoming [WebGPU](https://gpuweb.github.io/gpuweb/) standard. At the moment we've made good progress on a Python [wrapper](https://github.com/pygfx/wgpu-py). You can pip-install and try it!

We're also building a higher level render engine on top: https://github.com/pygfx/pygfx which is similar to / inspired by ThreeJS. It's very much in development, but eventually it could be a basis to build a high-level viz library on.

I have so much more to tell, but I don't want to hijack this thread more than I already did. I plan on writing some blog-posts on these new things. I'm happy to discuss more (in a new issue), e.g. about how this tech can play a role in the vispy community.

**Comment by sumanthratna** [comment]
Status update:

[`vispy/ext/cocoapy.py`](https://github.com/vispy/vispy/blob/master/vispy/ext/cocoapy.py) calls `cdll.LoadLibrary(util.find_library('...'))` a bunch of times, but ctypes can't find these libraries on Big Sur. On Catalina, `util.find_library('objc')` returns `/usr/lib/libobjc.dylib` and on Big Sur, `util.find_library('objc')` returns `None`. These are all the libraries loaded in `cocoapy.py`:
- objc
- CoreFoundation
- AppKit
- quartz
- CoreText
- Foundation

https://docs.python.org/3/library/ctypes.html#finding-shared-libraries
https://docs.python.org/3/library/ctypes.html#ctypes.LibraryLoader.LoadLibrary

**Comment by MrBenGriffin** [comment]
unfortunately vispy now seems to be dead in the water on MacOs.

**Comment by djhoese** [comment]
Would you mind providing more information at why you think this? Judging by the number of comments on #1885 I think people are making it work just fine (since the issue was a linking problem not a deprecation of OpenGL). VisPy devs (past and present) are exploring other non-OpenGL options for the future as well.

**Comment by MrBenGriffin** [comment]
@djhoese I revised my original comment - import is still not working on 3.8.7 Currently I'm having to use mayavi for visualisations which is okay, but I would prefer to use vispy.

**Comment by djhoese** [comment]
@MrBenGriffin Are you able to move to Python 3.9? As mentioned in #1885 most releases of Python 3.9 (the newer ones) should work on MacOS.

As for your revised comment, this is exactly the issue that #1885 is talking about. Please see all of the commentary there. The bottom line is that we have to wait for a Python 3.8 backport for the fixes that are now available in Python 3.9.

**Comment by MrBenGriffin** [comment]
@djhoese, I really apologise for the randomness - I'm right against a deadline, and I have a policy of reporting bugs when I can (being a developer, I know how useful they can be at times).  Thanks for pointing out #1885 - three times!  I will come back when I can use Python 3.9 or above.
